### DISCUSSION TITLE
Reenable a disabled query

### DISCUSSION THREAD

**Discussion by Gilthans** [TanStack/query#3701] [Q&A] [answered]
Hello,
I have a query that requires data from an external library to be resolved before it can be run.
I found two solutions for this:
 1) Create the query with `enable: false`, and then `useEffect` once the value is resolved to manually `refetch`. This isn't great because the query is still technically disabled. I could use `fetch({enabled: true})`, which looking at the code would re-enable the query, but that does not quite look intended? It's definitely not intuitive. (I'm wondering why setOptions is private?)
 2) Another option is to check for the resolved value inside the query, and if it's not yet resolved, return a sentinel value. This isn't ideal however, since I would like the query to be in a loading/idle state until the value is resolved, and doing this makes the query appear to have succeeded.

What is the correct approach here?

**Comment by TkDodo** [comment, marked as answer]
given that you manage the data from the external library within react (e.g. useState, some global client state manager, or even react-query if that data comes from an api call), the best way is to set `enabled` to a dynamic value, so that the query then enables automatically once the data comes in. So you're really looking at a standard [dependent query](https://react-query.tanstack.com/guides/dependent-queries):

```
const dataFromExternalLib = useDataFromExternalLib()

const { data } = useQuery(
  ['something', dataFromExternalLib],
  () => fetchSomething(dataFromExternalLib),
  {
    enabled: !!dataFromExternalLib
  }
)
```

**Reply by Gilthans** [reply]
That's interesting - and it did work, thanks!
I don't quite understand how that works though. How does the enable change get triggered?

**Reply by TkDodo** [reply]
if `dataFromExternalLib` is `undefined` or any other falsy value, `enabled` is `false`. If `dataFromExternalLib` then changes to a truthy value, and a re-render occurs, react-query will see `enabled: true` and will start to run the query.
